src/Messages.jl

```diff
@@ -0,0 +1,555 @@
+"""
+The `Messages` module defines structs that represent [`HTTP.Request`](@ref)
+and [`HTTP.Response`](@ref) Messages.
+
+The `Response` struct has a `request` field that points to the corresponding
+`Request`; and the `Request` struct has a `response` field.
+The `Request` struct also has a `parent` field that points to a `Response`
+in the case of HTTP Redirect.
+
+
+The Messages module defines `IO` `read` and `write` methods for Messages
+but it does not deal with URIs, creating connections, or executing requests.
+
+The `read` methods throw `EOFError` exceptions if input data is incomplete.
+and call parser functions that may throw `HTTP.ParsingError` exceptions.
+The `read` and `write` methods may also result in low level `IO` exceptions.
+
+
+### Sending Messages
+
+Messages are formatted and written to an `IO` stream by
+[`Base.write(::IO,::HTTP.Messages.Message)`](@ref) and or
+[`HTTP.Messages.writeheaders`](@ref).
+
+
+### Receiving Messages
+
+Messages are parsed from `IO` stream data by
+[`HTTP.Messages.readheaders`](@ref).
+This function calls [`HTTP.Messages.appendheader`](@ref) and
+[`HTTP.Messages.readstartline!`](@ref).
+
+The `read` methods rely on [`HTTP.IOExtras.unread!`](@ref) to push excess
+data back to the input stream.
+
+
+### Headers
+
+Headers are represented by `Vector{Pair{String,String}}`. As compared to
+`Dict{String,String}` this allows [repeated header fields and preservation of
+order](https://tools.ietf.org/html/rfc7230#section-3.2.2).
+
+Header values can be accessed by name using
+[`HTTP.Messages.header`](@ref) and
+[`HTTP.Messages.setheader`](@ref) (case-insensitive).
+
+The [`HTTP.Messages.appendheader`](@ref) function handles combining
+multi-line values, repeated header fields and special handling of
+multiple `Set-Cookie` headers.
+
+### Bodies
+
+The `HTTP.Message` structs represent the Message Body as `Vector{UInt8}`.
+
+Streaming of request and response bodies is handled by the
+[`HTTP.StreamLayer`](@ref) and the [`HTTP.Stream`](@ref) `<: IO` stream.
+"""
+
+
+module Messages
+
+export Message, Request, Response,
+       reset!,
+       iserror, isredirect, ischunked, issafe, isidempotent,
+       header, hasheader, setheader, defaultheader, appendheader,
+       mkheaders, readheaders, headerscomplete, readtrailers, writeheaders,
+       readstartline!, writestartline,
+       bodylength, unknown_length
+
+if VERSION > v"0.7.0-DEV.2338"
+using Unicode
+end
+
+import ..HTTP
```

Comment:
why is this needed?

Reply:
From memory it is to keep Documenter.jl happy. I think otherwise it fails with unknown var `HTTP` when processing doc strings that refer to fully qualified type names.